File: apps/website/src/optimizer.ts
import { calculateSimpleDamage } from './damage'
import {
  getNaturesByStatChanges,
  type Nature,
} from './natures'
import {
  calculateAttack,
  calculateDefense,
  calculateHp,
  calculateSpecialAttack,
  calculateSpecialDefense,
  calculateSpeed,
} from './stats'

export type HeldItem =
  | 'none'
  | 'eviolite'
  | 'assault_vest'

export interface PokemonBaseStats {
  base_hp: number
  base_atk: number
  base_def: number
  base_spa: number
  base_spd: number
  base_spe: number
}

export interface DefensiveStats {
  defense: number
  specialDefense: number
}

export interface BestDefensiveSpread {
  nature: Nature

  hp_points: number
  atk_points: number
  def_points: number
  spa_points: number
  spd_points: number
  spe_points: number

  hp: number
  attack: number
  raw_defense: number
  raw_special_defense: number
  defense: number
  special_defense: number
  special_attack: number
  speed: number

  defense_stage: number
  special_defense_stage: number
  held_item: HeldItem

  score: number
}

function validateStatStage(stage: number): void {
  if (
    !Number.isInteger(stage)
    || stage < -6
    || stage > 6
  ) {
    throw new RangeError(
      `Stat stage must be an integer from -6 to +6, received ${stage}`,
    )
  }
}

export function applyStatStage(
  stat: number,
  stage: number,
): number {
  validateStatStage(stage)

  if (stage >= 0) {
    return Math.floor(
      (stat * (2 + stage)) / 2,
    )
  }

  return Math.floor(
    (stat * 2) / (2 - stage),
  )
}

export function applyDefensiveItem(
  defense: number,
  specialDefense: number,
  heldItem: HeldItem,
): DefensiveStats {
  switch (heldItem) {
    case 'eviolite':
      return {
        defense: Math.floor(defense * 1.5),
        specialDefense: Math.floor(
          specialDefense * 1.5,
        ),
      }

    case 'assault_vest':
      return {
        defense,
        specialDefense: Math.floor(
          specialDefense * 1.5,
        ),
      }

    case 'none':
      return {
        defense,
        specialDefense,
      }

    default: {
      const exhaustiveCheck: never = heldItem

      throw new Error(
        `Unsupported held item: ${exhaustiveCheck}`,
      )
    }
  }
}

export function findBestDefensiveSpread(
  pokemon: PokemonBaseStats,
  increasedNatureStat: string,
  decreasedNatureStat: string,
  fixedAttackPoints = 0,
  fixedSpecialAttackPoints = 0,
  fixedSpeedPoints = 0,
  defenseStage = 0,
  specialDefenseStage = 0,
  heldItem: HeldItem = 'none',
): BestDefensiveSpread | null {
  const remainingPoints =
    66
    - fixedAttackPoints
    - fixedSpecialAttackPoints
    - fixedSpeedPoints

  let bestScore: number | null = null
  let bestSpread: BestDefensiveSpread | null = null

  const possibleNatures =
    getNaturesByStatChanges(
      increasedNatureStat,
      decreasedNatureStat,
    )

  const natureEntries = Object.values(
    possibleNatures,
  ).filter(
    (nature): nature is Nature =>
      nature !== undefined,
  )

  if (natureEntries.length === 0) {
    return null
  }

  for (const nature of natureEntries) {
    for (
      let hpPoints = 0;
      hpPoints <= 32;
      hpPoints += 1
    ) {
      for (
        let defensePoints = 0;
        defensePoints <= 32;
        defensePoints += 1
      ) {
        for (
          let specialDefensePoints = 0;
          specialDefensePoints <= 32;
          specialDefensePoints += 1
        ) {
          const defensivePointTotal =
            hpPoints
            + defensePoints
            + specialDefensePoints

          if (
            defensivePointTotal
            !== remainingPoints
          ) {
            continue
          }

          const hp = calculateHp(
            pokemon.base_hp,
            hpPoints,
          )

          const attack = calculateAttack(
            pokemon.base_atk,
            fixedAttackPoints,
            nature.attack,
          )

          const rawDefense = calculateDefense(
            pokemon.base_def,
            defensePoints,
            nature.defense,
          )

          const specialAttack =
            calculateSpecialAttack(
              pokemon.base_spa,
              fixedSpecialAttackPoints,
              nature.special_attack,
            )

          const rawSpecialDefense =
            calculateSpecialDefense(
              pokemon.base_spd,
              specialDefensePoints,
              nature.special_defense,
            )

          const speed = calculateSpeed(
            pokemon.base_spe,
            fixedSpeedPoints,
            nature.speed,
          )

          const stagedDefense = applyStatStage(
            rawDefense,
            defenseStage,
          )

          const stagedSpecialDefense =
            applyStatStage(
              rawSpecialDefense,
              specialDefenseStage,
            )

          const effectiveStats =
            applyDefensiveItem(
              stagedDefense,
              stagedSpecialDefense,
              heldItem,
            )

          const physicalDamage =
            calculateSimpleDamage(
              effectiveStats.defense,
            )

          const specialDamage =
            calculateSimpleDamage(
              effectiveStats.specialDefense,
            )

          const score =
            physicalDamage / hp
            + specialDamage / hp

          if (
            bestScore === null
            || score < bestScore
          ) {
            bestScore = score

            bestSpread = {
              nature,

              hp_points: hpPoints,
              atk_points: fixedAttackPoints,
              def_points: defensePoints,
              spa_points:
                fixedSpecialAttackPoints,
              spd_points:
                specialDefensePoints,
              spe_points: fixedSpeedPoints,

              hp,
              attack,

              raw_defense: rawDefense,
              raw_special_defense:
                rawSpecialDefense,

              defense: effectiveStats.defense,
              special_defense:
                effectiveStats.specialDefense,

              special_attack: specialAttack,
              speed,

              defense_stage: defenseStage,
              special_defense_stage:
                specialDefenseStage,
              held_item: heldItem,

              score,
            }
          }
        }
      }
    }
  }

  return bestSpread
}

File: apps/desktop/tools/defensive_spread_optimizer/optimizer.py
from .stats import (
    calculate_hp,
    calculate_attack,
    calculate_defense,
    calculate_special_attack,
    calculate_special_defense,
    calculate_speed
)

from .damage import calculate_simple_damage
from .natures import get_natures_by_stat_changes

VALID_ITEMS = {
    "none",
    "eviolite",
    "assault_vest",
}


def apply_stat_stage(stat: int, stage: int) -> int:
    """
    Apply a Pokémon stat stage from -6 to +6.

    Examples:
    +1 = x1.5
    +2 = x2
    -1 = x2/3
    -2 = x1/2
    """
    if not -6 <= stage <= 6:
        raise ValueError("Stat stage must be between -6 and +6.")

    if stage >= 0:
        return stat * (2 + stage) // 2

    return stat * 2 // (2 - stage)


def apply_defensive_item(
    defense: int,
    special_defense: int,
    held_item: str
) -> tuple[int, int]:
    """
    Apply the defensive effect of the selected held item.

    Eviolite:
        Defense and Special Defense x1.5

    Assault Vest:
        Special Defense x1.5
    """
    if held_item not in VALID_ITEMS:
        raise ValueError(f"Unknown held item: {held_item}")

    if held_item == "eviolite":
        defense = defense * 3 // 2
        special_defense = special_defense * 3 // 2

    elif held_item == "assault_vest":
        special_defense = special_defense * 3 // 2

    return defense, special_defense


def find_best_defensive_spread(
    pokemon,
    increased_nature_stat,
    decreased_nature_stat,
    fixed_atk_points=0,
    fixed_spa_points=0,
    fixed_spe_points=0,
    defense_stage=0,
    special_defense_stage=0,
    held_item="none"
):
    remaining_points = (
        66
        - fixed_atk_points
        - fixed_spa_points
        - fixed_spe_points
    )

    best_score = None
    best_spread = None

    possible_natures = get_natures_by_stat_changes(
        increased_nature_stat,
        decreased_nature_stat
    )

    if not possible_natures:
        return None

    for nature_name, nature in possible_natures.items():

        for hp_points in range(33):
            for def_points in range(33):
                for spd_points in range(33):

                    if (
                        hp_points
                        + def_points
                        + spd_points
                        != remaining_points
                    ):
                        continue

                    hp = calculate_hp(
                        pokemon["base_hp"],
                        hp_points
                    )

                    attack = calculate_attack(
                        pokemon["base_atk"],
                        fixed_atk_points,
                        nature["attack"]
                    )

                    raw_defense = calculate_defense(
                        pokemon["base_def"],
                        def_points,
                        nature["defense"]
                    )

                    special_attack = calculate_special_attack(
                        pokemon["base_spa"],
                        fixed_spa_points,
                        nature["special_attack"]
                    )

                    raw_special_defense = calculate_special_defense(
                        pokemon["base_spd"],
                        spd_points,
                        nature["special_defense"]
                    )

                    speed = calculate_speed(
                        pokemon["base_spe"],
                        fixed_spe_points,
                        nature["speed"]
                    )

                    # Apply stat stages first.
                    defense = apply_stat_stage(
                        raw_defense,
                        defense_stage
                    )

                    special_defense = apply_stat_stage(
                        raw_special_defense,
                        special_defense_stage
                    )

                    # Apply Eviolite or Assault Vest afterwards.
                    defense, special_defense = apply_defensive_item(
                        defense,
                        special_defense,
                        held_item
                    )

                    physical_damage = calculate_simple_damage(
                        defense
                    )

                    special_damage = calculate_simple_damage(
                        special_defense
                    )

                    score = (
                        physical_damage / hp
                        + special_damage / hp
                    )

                    if best_score is None or score < best_score:
                        best_score = score

                        best_spread = {
                            "nature": nature,

                            "hp_points": hp_points,
                            "atk_points": fixed_atk_points,
                            "def_points": def_points,
                            "spa_points": fixed_spa_points,
                            "spd_points": spd_points,
                            "spe_points": fixed_spe_points,

                            "hp": hp,
                            "attack": attack,

                            # Unmodified calculated stats
                            "raw_defense": raw_defense,
                            "raw_special_defense": (
                                raw_special_defense
                            ),

                            # Effective battle stats
                            "defense": defense,
                            "special_defense": special_defense,

                            "special_attack": special_attack,
                            "speed": speed,

                            "defense_stage": defense_stage,
                            "special_defense_stage": (
                                special_defense_stage
                            ),
                            "held_item": held_item,

                            "score": score
                        }

    return best_spread